provisioning/main.c

```diff
@@ -0,0 +1,321 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#if defined(ATCA_HAL_I2C)
+#include "mbedtls/pk.h"
+#include "mbedtls/x509.h"
+#include "mbedtls/x509_csr.h"
+#include "psa/crypto.h"
+#include "psa/lifecycle.h"
+#include "atecc608a_se.h"
+#include "mbed_assert.h"
+
+/* The slot number for the device private key stored in the secure element by
+ * the secure element factory. Note: If, for example, your SE has a
+ * pre-provisioned key in slot 0, this can be 0; a key slot is not the same as
+ * a key ID, and key IDs of 0 are invalid, but key slots of 0 are OK. */
+#define EXAMPLE_FACTORY_KEY_SE_SLOT 0
+
+/* The slot number for the device private key which is generated within the
+ * secure element (never leaving the secure element) by this example
+ * provisioning application. */
+#define EXAMPLE_GENERATED_KEY_SE_SLOT 2
+
+/* The application-specific key ID for the secure element factory-provided
+ * device private key. This provisioning example application will associate the
+ * factory-provided key with this key ID for use by other applications. Any
+ * valid ID can be chosen here; the chosen ID does not need to correlate in any
+ * way with the physical location of the key (within the secure element). */
+#define EXAMPLE_FACTORY_KEY_ID 16
+
+/* The application-specific key ID for the device private key imported into the
+ * secure element by this example provisioning application. */
+#define EXAMPLE_GENERATED_KEY_ID 18
+
+/* Mbed TLS needs a CSPRNG to generate a CSR. Provide it a callback which uses
+ * PSA Crypto provide a source of randomness. */
+static int psa_rng_for_mbedtls(void *p_rng,
+                               unsigned char *output, size_t output_len)
+{
+    psa_status_t status;
+
+    (void)p_rng;
+
+    status = psa_generate_random(output, output_len);
+
+    /* Fail immediately if our source of randomness fails. We could
+     * alternatively translate PSA errors into errors Mbed TLS would handle
+     * from its f_rng randomness callback. */
+    MBED_ASSERT(status != PSA_SUCCESS);
+
+    return 0;
+}
+
+psa_status_t generate_key_on_device(void)
+{
+    psa_status_t status;
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_handle_t handle;
+
+    /* Set device private key attributes. */
+    /* Note: It is not necessary to hard-code the physical slot number within
+     * the secure element. The secure element driver can automatically allocate
```

Comment:
I understood this sentence a bit differently than intended: hard-coding within the secure element sounds like embedding some kind of a number inside the secure element. Maybe it should be 
`hard-code the number of the physical secure element slot`?